core/sequencer/server.go

```diff
@@ -313,14 +315,24 @@ func (s *Server) ApplyRevision(ctx context.Context, in *spb.ApplyRevisionRequest
 		return nil, err
 	}
 
-	// Apply mutations to values.
-	newLeaves, err := runner.ApplyMutations(entry.ReduceFn, indexedValues, leaves,
-		func(err error) { glog.Warning(err); mutationFailures.Inc(err.Error()) },
-	)
+	// Map Map Leaves
+	indexedLeaves, err := runner.DoMapMapLeafFn(mapper.MapMapLeafFn, leaves)
 	if err != nil {
 		return nil, err
 	}
 
+	// GroupByIndex.
+	joined := runner.Join(indexedLeaves, indexedValues)
+
+	// Apply mutations to values.
+	newIndexedLeaves := runner.DoReduceFn(entry.ReduceFn, joined,
+		func(err error) { glog.Warning(err); mutationFailures.Inc(err.Error()) },
```

Comment:
This can be emitErrFn. But it strikes me we may want different counters for the different stages of the pipeline?

Reply:
Thanks for that catch. 

Yes, different errors for different stages of the pipeline - up next! (with metrics too)